plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php

```diff
@@ -0,0 +1,128 @@
+	 * Initialize the checkout link service.
+	 */
+	public function init() {
+		add_action( 'init', array( $this, 'add_checkout_link_endpoint' ) );
+		add_filter( 'query_vars', array( $this, 'add_query_vars' ), 0 );
+		add_action( 'template_redirect', array( $this, 'handle_checkout_link_endpoint' ) );
+	}
+
+	/**
+	 * Add the checkout link endpoint.
+	 */
+	public function add_checkout_link_endpoint() {
+		add_rewrite_rule( '^checkout-link$', 'index.php?checkout-link=true', 'top' );
+	}
+
+	/**
+	 * Add the checkout link query var.
+	 *
+	 * @param array $vars The query vars.
+	 * @return array The query vars.
+	 */
+	public function add_query_vars( $vars ) {
+		$vars[] = 'checkout-link';
+		return $vars;
+	}
+
+	/**
+	 * Handle the checkout link endpoint.
+	 *
+	 * @return void
+	 */
+	public function handle_checkout_link_endpoint() {
+		if ( ! get_query_var( 'checkout-link' ) ) {
+			return;
+		}
+		wp_safe_redirect( $this->get_checkout_link() );
+		exit;
+	}
+
+	/**
+	 * Process the query params and return the checkout link to redirect to complete with session token.
+	 *
+	 * @return string The checkout link.
+	 */
+	protected function get_checkout_link() {
+		$controller = new CartController();
+		$controller->empty_cart();
+
+		// Populate cart with products.
+		$products = array_filter( explode( ',', wc_clean( wp_unslash( $_GET['products'] ?? '' ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		foreach ( $products as $product_id_qty ) {
+			if ( strpos( $product_id_qty, ':' ) !== false ) {
+				list( $product_id, $qty ) = explode( ':', $product_id_qty );
+			} else {
+				$product_id = $product_id_qty;
+				$qty        = 1;
+			}
+			if ( ! absint( $product_id ) ) {
+				continue;
+			}
+			try {
+				$controller->add_to_cart(
+					[
+						'id'       => absint( $product_id ),
+						'quantity' => absint( $qty ),
+					]
+				);
+			} catch ( \Exception $e ) {
+				wc_add_notice( $e->getMessage(), 'error' );
+			}
+		}
+
+		// Apply coupon if provided.
+		$coupon = wc_format_coupon_code( wp_unslash( $_GET['coupon'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+
+		if ( wc_coupons_enabled() && ! empty( $coupon ) ) {
+			try {
+				$controller->apply_coupon( $coupon );
+			} catch ( \Exception $e ) {
+				wc_add_notice( $e->getMessage(), 'error' );
+			}
+		}
+
+		// Nothing was added to the cart. We need to redirect to the cart page with an error notice. Since guests may not
+		// have a session, add the notice in the query string.
+		if ( wc()->cart->is_empty() ) {
+			$empty_cart_notice = __( 'The provided checkout link was out of date or invalid. No products were added to the cart.', 'woocommerce' );
+			wc_add_notice( $empty_cart_notice, 'error' );
+
+			if ( ! wc()->session->has_session() ) {
+				return add_query_arg( 'wc_error', $empty_cart_notice, wc_get_cart_url() );
+			}
+
+			return wc_get_cart_url();
+		}
```

Comment:
At some point I noticed several notices, likely left over from not clearing up the session? 
<img width="1357" alt="image" src="https://github.com/user-attachments/assets/e93015a0-77d5-49c3-928f-5096effaef38" />

Reply:
I think its due to no session. https://github.com/woocommerce/woocommerce/pull/58140/commits/fe7bdda6e6901bc589be0ed69a1abd01e0442867 should clear it up.